trin-core/src/portalnet/overlay.rs

```diff
@@ -0,0 +1,169 @@
+use crate::portalnet::types::SszEnr;
+use crate::portalnet::{Enr, U256};
+use crate::utils::xor_two_values;
+
+use discv5::enr::NodeId;
+use discv5::kbucket::{Filter, KBucketsTable};
+use parking_lot::RwLock;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Maximum number of ENRs in response to FindNodes.
+const FIND_NODES_MAX_NODES: usize = 32;
+/// Maximum numer of ENRs in response to FindContent.
+const FIND_CONTENT_MAX_NODES: usize = 32;
+
+#[derive(Clone)]
+pub struct Node {
+    enr: Enr,
+    data_radius: U256,
+}
+
+impl Node {
+    pub fn enr(&self) -> Enr {
+        self.enr.clone()
+    }
+
+    pub fn data_radius(&self) -> U256 {
+        self.data_radius.clone()
+    }
+}
+
+impl std::cmp::Eq for Node {}
+
+impl PartialEq for Node {
+    fn eq(&self, other: &Self) -> bool {
+        self.enr == other.enr
+    }
+}
+
+/// Configuration parameters for the overlay network.
+#[derive(Clone)]
+pub struct Config {
+    bucket_pending_timeout: Duration,
+    max_incoming_per_bucket: usize,
+    table_filter: Option<Box<dyn Filter<Node>>>,
+    bucket_filter: Option<Box<dyn Filter<Node>>>,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            bucket_pending_timeout: Duration::from_secs(60),
+            max_incoming_per_bucket: 16,
+            table_filter: None,
+            bucket_filter: None,
+        }
+    }
+}
+
+/// The node state for a node in an overlay network on top of Discovery v5.
+pub struct Overlay {
+    // The ENR of the local node.
+    local_enr: Arc<RwLock<Enr>>,
+    // The data radius of the local node.
+    data_radius: Arc<RwLock<U256>>,
```

Comment:
Does anything need to happen if either of these update? Like sending an announcement? Could we worth looking at the broadcast or watch channels in tokio.

Reply:
For `data_radius`, I don't think we would require a channel because `Overlay` would encapsulate all logic around `data_radius`.

For `local_enr`, I think those updates would originate from `Discv5`, which has public methods to update the ENR.